### ISSUE TITLE
Data for hg38

### ISSUE THREAD

**Issue by fridells51** [ZjGaothu/EpiGePT#5]
Hi, I am interested in using this model, but I'm finding it difficult to follow some of the reproduction section on the github. The reproduction section mentions a script called "Download_raw_data.sh," but this seems to be defunct. It has functionality for hg19 (while other parts of the tutorials have been updated for hg38) and in the current version on github, the script "download_raw_data.sh" relies on some file `/data/encode/metadata.tsv', but this file does not exist on the current repository.

On the download page, (https://health.tsinghua.edu.cn/epigept/download.php) the github indicates that this contains the processed training data for hg38. I assume it is the ~100-gigabyte file under "Target Data", `targets_data_v1.npy`. This file has shape (105, 15870000, 8). 

I assume this is 105 hg38 cell types (as detailed in `/data/encode/celltypes_and_tissues_hg38.txt`), 8 epigenomic tracks, and 15,870,000 bins of 128 bp across the entire genome. I have 2 questions:

1. how can I find the coordinates of the 128 bp bins? Is there any bed file containing the coordinates/intervals of these bins in the hg38 reference genome? can this be inferred from the "location_map" files in the `reference_motifscore` from the download page? 

2. Is there a working version of the download and preprocessing scripts for the hg38 data?

**Comment by ZjGaothu** [comment]
Hi, thank you for your recognition of our work and for your thoughtful questions.

Regarding your first question, we indeed overlooked providing the coordinates of the regions, and these cannot be inferred from the “location_map” files in the `reference_motifscore`. We have now uploaded the file containing the coordinates of the 128 bp bins to GitHub:  
👉 [regions_hg38.128k.bed](https://github.com/ZjGaothu/EpiGePT/blob/main/data/regions_hg38.128k.bed)

Each row in this file corresponds to a 128 kbp region, which includes 1,000 bins of 128 bp. There are a total of 15,870 long regions in this file.

For the download and preprocessing of hg38 data, the procedure is essentially the same as the one we provided for hg19.  
We have uploaded the metadata used for hg38 to the EpiGePT website:  
👉 [metadata_hg38.tsv](http://health.tsinghua.edu.cn/epigept/help/metadata_hg38.tsv)

You can download the original data using this metadata file, and modify the genome reference filter to hg38 during the download step.

Thank you again for your interest and questions. I hope this helps!

**Comment by fridells51** [comment]
Thanks so much for this update. One more thing, i noticed for 'train_idx_5_fold.npy' and 'test_idx_5_fold.npy', these files may correspond to the hg19 data. Unless I am mistaken, I notice that they have indices ranging from 0-128, which is consistent with the 129 ENCODE biosamples that were used to train the hg19 model. Do similar test and train indices exist for the hg38 model? I am trying to simply replicate the cross-region, cross-type, and cross-both result in Fig 1. 

I have downloaded the target data (targets_data_v1.npy), motifdata and reference transcription factor profiles from the download page, I have the hg38 reference and the bed file that you've provided above, and I have working environments for running the model. I believe all that I would need is to know what the indices for held-out cell types and held-out regions are. The manuscript mentions 19 unseen cellular contexts and chromosomes 18, 19, X, and Y as the unseen regions. Are these the held-out genomic regions? Do you have any information on what the 19 cell types that are withheld from the training data are?

**Comment by ZjGaothu** [comment]
Sorry for the delayed response. Regarding the hg38 data, we performed a random partition of the training and testing cell types. The indices for the training and testing sets are available at [train_cells.npy](https://github.com/ZjGaothu/EpiGePT/blob/main/model_hg38/train_cells.npy) and [test_cells.npy](https://github.com/ZjGaothu/EpiGePT/blob/main/model_hg38/test_cells.npy), respectively. We performed the evaluation on the subset of test cell line/tissues that have available epigenomic profiles. In addition, 13 cell lines/tissues can be matched with the chromatin interaction data from HiChIPdb, and their indices are [1, 8, 14, 21, 26, 38, 41, 49, 57, 61, 70, 75, 89]. These indices can also be included in the training set. Since we did not perform any cross‑region experiments for this subset, all available regions were used for training.

Specifically, the results for the cross‑region, cross‑cell‑type, and cross‑both settings in the hg38 genome were not presented in Figure 2. These settings were only evaluated on the DNase‑seq signal in the hg19 data.

I hope this clarifies your question.

**Comment by fridells51** [comment]
Just to be clear, these indices are 1-based, not 0-based, right? The tf expression file has 105 columns named 1,2,3,...,105. These "indices" are the column names that correspond to the hg38 tissues/cell types? 1 = A549, 2 = A673, ... 105 = vagina

**Comment by fridells51** [comment]
I'm specifically trying to replicate Fig S8 panel c/d from the supplementary information. 

This is how i load the model: 
```
from model_hg38 import EpiGePT
from model_hg38.config import *

model = EpiGePT.EpiGePT(WORD_NUM, TF_DIM, BATCH_SIZE)
checkpoint = torch.load(model_path, map_location=device)
model.load_state_dict(checkpoint['state_dict'])
model.eval()
```

and this is the inference workflow I use:

```
seq = self.genome[chrom][start:end]
seq_matrix = self._seq2mat(seq)  # ACGT → 4-channel one-hot encoding

motif_scores = self.motif_cache.get_motif_scores(chrom, start, end, tf_list, motif2tf, isContain)
tf_features = motif_scores.astype(np.float32) * tf_expression_log  # Element-wise multiplication
tf_features_padded = np.pad(tf_features, ((0, 0), (0, 1)), 'constant')  # Pad to 712 dims

seq_tensor = torch.FloatTensor(seq_matrix).unsqueeze(0).to(device)
tf_tensor = torch.FloatTensor(tf_features_padded).unsqueeze(0).to(device)

with torch.no_grad():
    prediction = self.model(seq_tensor, tf_tensor)  # Same model call as tutorial
    prediction_np = prediction.cpu().detach().numpy().squeeze(0)
```

These are the inputs I use that were all downloaded from the download page:
Pretrained model: `pretrain_models_hg38/pretrained_epigept.ckpt`
Reference genome: `hg38.fa` 
TF expression data: `tf_expression.csv `
Test cell types: `model_hg38/test_cells.npy` (i moved this here, but it's the same file you referenced above, downloaded from github)
Evaluation regions: `regions_hg38.128k.bed` (downloaded from github)
Motif scores: Pre-computed motif scanning results in `/reference_motifscore/` 

My main question is about the TF expression data. This file `tf_expression.csv` is part of the .zip file in the tf download page, but I haven't quantile normalized it against the reference_TPM_expr.csv file. Should I? Is this the correct TF file to use for replication? 

I'm using the data in targets_data_v1.npy that i downloaded from the epigept download portal for correlations after I perform the replication. I noticed a few cell indices (`[33, 37, 45, 48, 54, 63, 78, 85, 92, 99]`) are masked out in this file. I'm only interested in the 19 held-out cell types in `test_cells.npy`

```
np.load('test_cells.npy')
array([ 19,  29,  30,  32,  36,  46,  52,  56,  59,  60,  64,  69,  82,
        86,  90,  94,  95,  97, 104])
```

since none of these indices overlap with the masked indices, I am assuming that the 19 values in test_cells.npy are indices, and not cell type IDs. The reason this is important to understand for me is that in the TF expression files, the columns are named 1-105 based on cell type IDs. However, in the `targets_data_v1.npy` file, the dimension for cell type is 0-based, as it is a python array. I need to know whether these values in `test_cells.npy` correspond to cell IDs, or 0-based indices of the first dimension of `targets_data_v1.npy`. If you could clarify that for me, it would be very helpful. Thank you!
